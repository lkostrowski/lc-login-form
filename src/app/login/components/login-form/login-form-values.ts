export interface LoginFormValues {
    email: string;
    password: string;
    keepLogged: boolean;
}
