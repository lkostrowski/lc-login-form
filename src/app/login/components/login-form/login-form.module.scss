.globalError {
    margin-bottom: 3rem;
}

.field {
    margin-bottom: 2rem;
}

.keepLogged {
    margin-bottom: 4rem;
}
