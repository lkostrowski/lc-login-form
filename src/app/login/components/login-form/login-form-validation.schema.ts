import { object, string } from 'yup';
import { LoginFormValues } from './login-form-values';

export const loginFormValidationSchema = object<LoginFormValues>({
    email: string()
        .email()
        .required(),
    password: string()
        .length(6)
        .test(
            'oneUpper',
            'Password must contain at least one uppercase letter',
            (value: string) => {
                return /[A-Z]/.test(value);
            },
        )
        .test(
            'oneLower',
            'Password must contain at least one lowercase letter',
            (value: string) => {
                return /[a-z]/.test(value);
            },
        )
        .test(
            'oneDigit',
            'Password must contain at least one number',
            (value: string) => {
                return /[0-9]/.test(value);
            },
        )
        .required(),
});
