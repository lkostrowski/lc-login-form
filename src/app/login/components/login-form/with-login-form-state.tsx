import { withFormik } from 'formik';

import { LoginFormOuterProps } from './login-form.component';
import { LoginFormValues } from './login-form-values';
import { loginFormValidationSchema } from '../../validators/login-form-validation.schema';

export const withLoginFormState = withFormik<
    LoginFormOuterProps,
    LoginFormValues
>({
    handleSubmit(values, formikBag) {
        formikBag.props.onFormSubmit(values.email, values.password);
    },
    mapPropsToValues(props) {
        return {
            email: '',
            password: '',
        };
    },
    validationSchema: loginFormValidationSchema,
    validateOnBlur: true,
    validateOnChange: true,
});
