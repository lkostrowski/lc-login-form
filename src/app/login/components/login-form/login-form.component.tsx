import React, { HTMLProps } from 'react';
import { FormikProps } from 'formik';
import cx from 'classnames';

import { LoginFormValues } from './login-form-values';
import { TextField } from '../../../common/text-field/text-field.component';
import { compose } from 'recompose';
import { withLoginFormState } from './with-login-form-state';
import { Button } from '../../../common/button/button.component';
import { ErrorBar } from '../../../common/error-bar/error-bar.component';

import { ReactComponent as ArrowIcon } from '../../../../images/arrow.svg';

import styles from './login-form.module.scss';

export interface LoginFormOuterProps extends HTMLProps<HTMLFormElement> {
    onFormSubmit(email: string, password: string): unknown;
    globalError?: string;
}

interface Props extends FormikProps<LoginFormValues>, LoginFormOuterProps {}

export const LoginForm: React.FC<Props> = ({
    values,
    errors,
    handleChange,
    handleSubmit,
    handleBlur,
    globalError,
    className,
    touched,
}) => {
    return (
        <form onSubmit={handleSubmit} className={cx(className)}>
            {globalError && (
                <ErrorBar className={styles.globalError}>
                    {globalError}
                </ErrorBar>
            )}
            <TextField
                errorMessage={errors.email}
                hasError={Boolean(errors.email && touched.email)}
                label="E-mail"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.email}
                name="email"
                className={styles.field}
            />
            <TextField
                hasError={Boolean(errors.password && touched.password)}
                label="Password"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.password}
                name="password"
                className={styles.field}
            />
            <Button fullWidth type="submit" rightIcon={<ArrowIcon />}>
                Let me in!
            </Button>
        </form>
    );
};

export const StatefulLoginForm = compose<Props, LoginFormOuterProps>(
    withLoginFormState,
)(LoginForm);
