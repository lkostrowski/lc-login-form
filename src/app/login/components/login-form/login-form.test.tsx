import React from 'react';
import { render, fireEvent, wait, cleanup } from '@testing-library/react';
import { FormikProps } from 'formik';
import { LoginFormValues } from './login-form-values';
import { LoginForm, StatefulLoginForm } from './login-form.component';
import nanoID from 'nanoid';

afterEach(cleanup);

jest.mock('nanoid');

describe('Login form component', () => {
    beforeEach(() => {
        (nanoID as jest.Mock)
            .mockImplementationOnce(() => 'id-1')
            .mockImplementationOnce(() => 'id-2');
    });

    describe('Snapshots', () => {
        const formikPropsMock: Partial<FormikProps<LoginFormValues>> = {
            errors: {},
            values: {
                email: '',
                password: '',
            },
            handleChange: () => {},
            handleBlur: () => {},
            handleSubmit: () => {},
        };

        it('Renders empty state', () => {
            const { container } = render(
                <LoginForm
                    {...formikPropsMock as FormikProps<LoginFormValues>}
                    onFormSubmit={() => {}}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('Renders filled state', () => {
            const { container } = render(
                <LoginForm
                    {...formikPropsMock as FormikProps<LoginFormValues>}
                    onFormSubmit={() => {}}
                    values={{
                        password: 'Asd123!',
                        email: 'lukasz@ostrowski.ninja',
                    }}
                />,
            );

            expect(container).toMatchSnapshot();
        });

        it('Renders field error state', () => {
            const { container } = render(
                <LoginForm
                    {...formikPropsMock as FormikProps<LoginFormValues>}
                    onFormSubmit={() => {}}
                    values={{
                        password: '',
                        email: 'lukasz@ostrowski.ninja',
                    }}
                    errors={{
                        password: 'Password is required',
                    }}
                />,
            );

            expect(container).toMatchSnapshot();
        });
    });

    describe('Integration', () => {
        it('Should let user fill and submit the form', async () => {
            const mockSubmit = jest.fn();

            const { getByLabelText, getByText } = render(
                <StatefulLoginForm onFormSubmit={mockSubmit} />,
            );

            fireEvent.input(getByLabelText('E-mail'), {
                target: {
                    value: 'lukasz@ostrowski.ninja',
                },
            });

            fireEvent.input(getByLabelText('Password'), {
                target: {
                    value: 'Asd123',
                },
            });

            await wait();

            fireEvent.click(getByText('Let me in!'));

            await wait();

            expect(mockSubmit).toBeCalled();
        });

        it('Should ignore submit if user doesnt fill form and show errors', async () => {
            const mockSubmit = jest.fn();

            const { getByText, getAllByTestId } = render(
                <StatefulLoginForm onFormSubmit={mockSubmit} />,
            );

            fireEvent.click(getByText('Let me in!'));

            await wait();

            expect(getAllByTestId('text-field:error-message')).toHaveLength(2);

            expect(mockSubmit).not.toBeCalled();
        });

        it('Should submit if errors are corrected', async () => {
            const mockSubmit = jest.fn();

            const { getByText, getByLabelText } = render(
                <StatefulLoginForm onFormSubmit={mockSubmit} />,
            );

            fireEvent.click(getByText('Let me in!'));

            await wait();

            expect(mockSubmit).not.toBeCalled();

            fireEvent.input(getByLabelText('E-mail'), {
                target: {
                    value: 'lukasz@ostrowski.ninja',
                },
            });

            fireEvent.input(getByLabelText('Password'), {
                target: {
                    value: 'Asd123',
                },
            });

            await wait();

            fireEvent.click(getByText('Let me in!'));

            await wait();

            expect(mockSubmit).toBeCalled();
        });
    });
});
