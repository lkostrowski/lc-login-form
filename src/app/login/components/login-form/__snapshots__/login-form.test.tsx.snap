// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form component Snapshots Renders empty state 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-1"
      >
        E-mail
      </label>
      <input
        class="field"
        id="id-1"
        name="email"
        value=""
      />
    </div>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-2"
      >
        Password
      </label>
      <input
        class="field"
        id="id-2"
        name="password"
        value=""
      />
    </div>
    <button
      class="button"
    >
      Let me in!
    </button>
  </form>
</div>
`;

exports[`Login form component Snapshots Renders field error state 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-1"
      >
        E-mail
      </label>
      <input
        class="field"
        id="id-1"
        name="email"
        value="lukasz@ostrowski.ninja"
      />
    </div>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-2"
      >
        Password
      </label>
      <input
        class="field"
        id="id-2"
        name="password"
        value=""
      />
      <span
        class="error"
        data-testid="text-field:error-message"
      >
        Password is required
      </span>
    </div>
    <button
      class="button"
    >
      Let me in!
    </button>
  </form>
</div>
`;

exports[`Login form component Snapshots Renders filled state 1`] = `
<div>
  <form>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-1"
      >
        E-mail
      </label>
      <input
        class="field"
        id="id-1"
        name="email"
        value="lukasz@ostrowski.ninja"
      />
    </div>
    <div
      class="container"
    >
      <label
        class="label"
        for="id-2"
      >
        Password
      </label>
      <input
        class="field"
        id="id-2"
        name="password"
        value="Asd123!"
      />
    </div>
    <button
      class="button"
    >
      Let me in!
    </button>
  </form>
</div>
`;
